### DISCUSSION TITLE
Unit testing & functional testing for LINK cartridges

### DISCUSSION THREAD

**Discussion by StelianFeciucOSF** [SalesforceCommerceCloud/sfcc-ci#451] [Team Posts]
Hello,

We are in the process of certificating a new LINK cartridge due to the current requirements, we need to provide unit and functional tests for the given cartridge. Are you aware what tools should be used and if there are any guides or documentation for the best practices that should be followed?

Thank you!

**Comment by zlatinz** [comment]
Hi @StelianFeciucOSF,
1. Check https://documentation.b2c.commercecloud.salesforce.com/DOC1/topic/com.demandware.dochelp/SFRA/TestingSFRA.html for the tools and technologies used.

2. Note that as the tests are not run in the context of the platform (they are run on node.js locally) you need to define mocks as part of your tests.
See https://github.com/SalesforceCommerceCloud/storefront-reference-architecture/tree/master/test/mocks
and https://github.com/SalesforceCommerceCloud/dw-api-mock
3. I would suggest looking at some of the existing tests, e.g. https://github.com/SalesforceCommerceCloud/storefront-reference-architecture/blob/master/test/unit/app_storefront_base/models/account.js
https://github.com/SalesforceCommerceCloud/storefront-reference-architecture/blob/master/test/unit/app_storefront_base/scripts/checkout/checkoutHelpers.js
4. If you want to be thorough, you are also able to test controllers, using mocha, chai and sinonJS. See https://scotch.io/tutorials/how-to-test-nodejs-apps-using-mocha-chai-and-sinonjs for some tutorial that is related to node.js, but the principle is similar.

@SGD1953 please can you also comment here?

**Comment by pravngaur** [comment]
Did you explore Mocha? SFRA is also shipped with mocha for test case
coverage.

**Comment by StelianFeciucOSF** [comment]
Hello @zlatinz ,
Thank you for the recommendation and all the documents and possible tools!

Hello @pravngaur , 
We didn't get the chance to explore Mocha yet, but given all the documents found and received, I think that we will proceed with Mocha further. Thank you!
